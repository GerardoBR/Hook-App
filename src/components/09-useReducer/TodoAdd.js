import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ {description }, handleInputChange , reset] =useForm({
        description: ''
    });
    const handleAddSumbit =(e)=>{
        e.preventDefault();
        if(description.trim().length<=1){
            return ;
        }
        const newTodo = {
                id : new Date().getTime(),
                desc : description,
                done : false
        }
        // const actionAddTodo ={
        //     type : 'add',
        //     payload : newTodo
        // }
        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
            <h4>Agregar To do</h4>
                    <hr/>

                    <form onSubmit = { handleAddSumbit }>
                        <input 
                            type="text"  
                            className = " form-control "
                            name = "description" 
                            placeholder = "tarea a hacer" 
                            value = { description }
                            onChange = { handleInputChange }
                        />
                        <button 
                            type ="submit"
                            className = "  form-control btn btn-outline-primary mt-1 btn-block"
                        > 
                            Agregar
                        </button>
                    </form>
        </>
    )
}
