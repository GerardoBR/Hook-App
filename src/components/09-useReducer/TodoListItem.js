import React from 'react'

export const TodoListItem = ( {todo , hanldeToggle , handleDelete ,index} ) => {
    return (
        <li 
            key ={ todo.id }
            className="list-group-item"
        > 
        <p 
            className={ `${todo.done  && 'complete'}` }
            onClick = { ()=>hanldeToggle(todo.id)}
        
        >{ index }. { todo.desc }</p>
            <button className= "btn btn-danger" onClick= { ()=>{ handleDelete(todo.id)  } }> Borrar</button>
        </li>
    )
}
