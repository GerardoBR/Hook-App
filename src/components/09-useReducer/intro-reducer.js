// console.log("hola mundo")
const initialSatet =[{
    id : 1,
    todo : 'comprar jamon',
    done : false
}];


const todoReducer =(state = initialSatet, action)=>{

    if ( action?.type === 'agregar'){
        return [...state,action.payload];
    }
    return state;
}
 let todos = todoReducer();

//  AÃ±adir un nuevo elemento al todo
// No usar push en react 
const newTodo ={
    id : 1,
    todo : 'comprar jugo',
    done : false
};

const addToDoAction ={
    type : 'agregar',
    payload : newTodo
}
todos = todoReducer( todos , addToDoAction);

 console.log(todos);
 