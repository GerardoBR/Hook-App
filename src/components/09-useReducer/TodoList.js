import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( {todos ,hanldeToggle , handleDelete} ) => {
    return (
        <div>
            <ul className="list-group">
                {
                    
                    todos.map( ( todo, i ) =>(
                      <TodoListItem
                        todo = { todo }
                        handleDelete = { handleDelete }
                        hanldeToggle = { hanldeToggle }
                        index = { i+1 }
                        key ={ todo.id }

                      />
                        
                    ))
                }
            </ul> 
        </div>
    )
}
