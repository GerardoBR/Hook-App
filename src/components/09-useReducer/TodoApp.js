import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './toReducer';
import './todo.css';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init =()=>{
    // return [{
    //     id : new Date().getTime(),
    //     desc : ' aprender React',
    //     done : false
    // }];
    return JSON.parse(localStorage.getItem('todos')) || [];

}
export const TodoApp = () => {
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    

    // console.log(description);
    useEffect(()=>{
        localStorage.setItem('todos' , JSON.stringify(todos));
    },[todos]);

    const handleDelete = (todoid)=>{
        const actionDelete={
            type: 'delete',
            payload : todoid
        }
        dispatch( actionDelete );
    }
    const hanldeToggle=(todoid)=>{

        dispatch({
            type: 'toggle',
            payload : todoid
        })
    }

    // const handleAddSumbit =(e)=>{
    //     e.preventDefault();
        
    //     const newTodo = {
    //         id : new Date().getTime(),
    //         desc : description,
    //         done : false
    //     }
    //     const actionAddTodo ={
    //         type : 'add',
    //         payload : newTodo
    //     }
    //     dispatch(actionAddTodo);
    //     reset();
    // }
    const handleAddTodo  = ( newTodo)=>{

        dispatch({
            type : 'add',
            payload : newTodo
        })
    }

    return (
        <div>
            <h1>TodoApp : ( { todos.length } )</h1>
            <hr/>
            <div className=" row">
                <div className = "col-6">
                    <TodoList 
                        todos = { todos }
                        hanldeToggle = { hanldeToggle }
                        handleDelete = { handleDelete }
                       
                    
                    />
                </div>
                <div className="col-6">
                    <TodoAdd 
                        handleAddTodo  = { handleAddTodo }
                    />
                </div>
                
            </div>
        </div>
    )
}


