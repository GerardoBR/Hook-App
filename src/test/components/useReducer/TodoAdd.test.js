import { shallow } from "enzyme"
import { TodoAdd } from "../../../components/09-useReducer/TodoAdd"

describe('Pruebas sobre <TodoAdd/>', () => {
    const handleAddTodo = jest.fn();
    const wrapper = shallow(
    <TodoAdd
        handleAddTodo ={ handleAddTodo }
    />
    )
    test('debe de mostrarse correctamente ', () => {
        
        expect ( wrapper ).toMatchSnapshot();
    });

    test('No debe de lllamar handleAddTodo', () => {
        
        const formSubmit = wrapper.find('form').prop('onSubmit');
        
        formSubmit({preventDefault(){}});
        expect( handleAddTodo ).toHaveBeenCalledTimes(0);

    });
    test('debe de llamar HandlAddTodo', () => {
            // Con argumento
        const value = "aprender EXPRESS"
        wrapper.find('input').simulate('change',{
            target:{
                value : value ,
                name : 'description'
            }
        });
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault(){}});
        expect( handleAddTodo ).toHaveBeenCalledTimes(1);
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) );
        expect( handleAddTodo ).toHaveBeenCalledWith( {
            id : expect.any(Number),
            desc: value,
            done : false
        });
        expect ( wrapper.find('input').prop('value') ).toBe('');
    })
    
    
    
})

