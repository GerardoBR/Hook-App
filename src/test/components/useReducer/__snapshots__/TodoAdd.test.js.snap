// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoAdd/> debe de mostrarse correctamente  1`] = `
<Fragment>
  <h4>
    Agregar To do
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      className=" form-control "
      name="description"
      onChange={[Function]}
      placeholder="tarea a hacer"
      type="text"
      value=""
    />
    <button
      className="  form-control btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
